dll/win32/shlwapi/autocomp.c

```diff
@@ -0,0 +1,707 @@
+                }
+
+                if (UrlIsW(szValue, URLIS_URL))
+                {
+                    AC_EnumString_AddString(pES, szValue);
+                }
+            }
+        }
+
+        RegCloseKey(hKey);
+    }
+}
+
+static HRESULT STDMETHODCALLTYPE
+AC_EnumString_Reset(IEnumString* This)
+{
+    DWORD attrs;
+    WCHAR szText[MAX_PATH];
+    AC_EnumString *this = (AC_EnumString *)This;
+    DWORD dwSHACF_ = this->m_dwSHACF_;
+
+    GetWindowTextW(this->m_hwndEdit, szText, ARRAYSIZE(szText));
+    attrs = GetFileAttributesW(szText);
+
+    AC_EnumString_ResetContent(this);
+
+    if (dwSHACF_ & (SHACF_FILESYS_ONLY | SHACF_FILESYSTEM | SHACF_FILESYS_DIRS))
+    {
+        BOOL bDirOnly = !!(dwSHACF_ & SHACF_FILESYS_DIRS);
+        if (attrs != INVALID_FILE_ATTRIBUTES)
+        {
+            if (attrs & FILE_ATTRIBUTE_DIRECTORY)
+            {
+                AC_DoDir(this, szText, bDirOnly);
+            }
+        }
+        else if (szText[0])
+        {
+            PathRemoveFileSpecW(szText);
+            AC_DoDir(this, szText, bDirOnly);
+        }
+        else
+        {
+            AC_DoFiles(this, bDirOnly);
+        }
+    }
+
+    if (!(dwSHACF_ & (SHACF_FILESYS_ONLY)))
+    {
+        if (dwSHACF_ & SHACF_URLHISTORY)
+        {
+            AC_DoURLHistory(this);
+        }
+        if (dwSHACF_ & SHACF_URLMRU)
+        {
+            AC_DoURLMRU(this);
+        }
+    }
+
+    this->m_istr = 0;
+    return S_OK;
+}
+
+static AC_EnumString *
+AC_EnumString_Construct(SIZE_T capacity)
+{
+    AC_EnumStringVtbl *lpVtbl;
+    AC_EnumString *ret = CoTaskMemAlloc(sizeof(AC_EnumString));
+    if (!ret)
+        return ret;
+
+    lpVtbl = CoTaskMemAlloc(sizeof(AC_EnumStringVtbl));
+    if (!lpVtbl)
+    {
+        CoTaskMemFree(ret);
+        return NULL;
+    }
+
+    lpVtbl->QueryInterface = AC_EnumString_QueryInterface;
+    lpVtbl->AddRef = AC_EnumString_AddRef;
+    lpVtbl->Release = AC_EnumString_Release;
+    lpVtbl->Next = AC_EnumString_Next;
+    lpVtbl->Skip = AC_EnumString_Skip;
+    lpVtbl->Reset = AC_EnumString_Reset;
+    lpVtbl->Clone = AC_EnumString_Clone;
+    ret->lpVtbl = lpVtbl;
+
+    ret->m_pstrs = (BSTR *)CoTaskMemAlloc(capacity * sizeof(BSTR));
+    if (ret->m_pstrs)
+    {
+        ret->m_cRefs = 1;
+        ret->m_istr = 0;
+        ret->m_cstrs = 0;
+        ret->m_capacity = capacity;
+        ret->m_hwndEdit = NULL;
+        return ret;
+    }
+
+    CoTaskMemFree(ret->lpVtbl);
+    CoTaskMemFree(ret);
+    return NULL;
+}
+
+static BOOL
+AC_AdaptFlags(HWND hwndEdit, LPDWORD pdwACO_, LPDWORD pdwSHACF_)
+{
+    static const LPCWSTR s_pszAutoComplete =
+        L"Software\\Microsoft\\Internet Explorer\\AutoComplete";
+
+    DWORD dwType, cbValue, dwACO_, dwSHACF_;
+    WCHAR szValue[8];
+
+    dwSHACF_ = *pdwSHACF_;
+
+    dwACO_ = 0;
+    if (dwSHACF_ == SHACF_DEFAULT)
+        dwSHACF_ = SHACF_FILESYSTEM | SHACF_URLALL;
+
+    if (dwSHACF_ & SHACF_AUTOAPPEND_FORCE_OFF)
+    {
+        ;
```

Comment:
What's this? Should it be implemented, but "later"? If so, please use `UNIMPLEMENTED;` instead.
If this case on the contrary should never be implemented, please add a little comment.